File: DMOJ/DMOPC/2017/'17 Contest 2/BadBunnies.java
// DMOPC '17 Contest 2 P3 - Bad Bunnies

import java.util.*;
import java.io.*;
public class BadBunnies {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException{
		int n = readInt(), r = readInt();
		ArrayList<Integer> adj[] = new ArrayList[n+1];
		for (int i=1; i<=n; i++) {
			adj[i] = new ArrayList<Integer>();
		}
		for (int i=1; i<n; i++) {
			int u = readInt(), v = readInt();
			adj[u].add(v); adj[v].add(u);
		}
		Queue<Integer> q = new LinkedList<>();
		boolean v[] = new boolean[n+1];
		int dis[] = new int[n+1];
		for (int i=1; i<=r; i++) {
			int a = readInt();
			q.add(a);
			v[a] = true;
			dis[a] = 0;
		}
		while (!q.isEmpty()) {
			int cur = q.poll();
			for (int nxt: adj[cur]) {
				if (!v[nxt]) {
					q.add(nxt);
					v[nxt] = true;
					dis[nxt] = dis[cur]+1;
				}
			}
		}
		int start = readInt(), end = readInt(), prev[] = new int[n+1];
		Arrays.fill(v, false);
		q.add(start);
		v[start] = true;
		prev[start] = -1;
		while (!q.isEmpty()) {
			int cur = q.poll();
			for (int nxt: adj[cur]) {
				if (!v[nxt]) {
					q.add(nxt);
					v[nxt] = true;
					prev[nxt] = cur;
				}
			}
		}
		int ans = Integer.MAX_VALUE;
		for (int i=end; i!=-1; i=prev[i]) {
			ans = Math.min(ans, dis[i]);
		}
		System.out.println(ans);
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: DMOJ/DMOPC/2017/'17 Contest 2/BadBunnies.cpp
// DMOPC '17 Contest 2 P3 - Bad Bunnies

#include <bits/stdc++.h>
using namespace std;
// clang-format off
#ifdef __linux__
#define getchar getchar_unlocked
#endif
#define su(x) do{while((x=getchar())<48); for(x-=48; 48<=(_=getchar()); x=(x<<3)+(x<<1)+_-48);}while(0)
#define si(x) do{while((x=getchar())<45); _sign=x==45; if(_sign) while((x=getchar())<48); for(x-=48; 48<=(_=getchar()); x=(x<<3)+(x<<1)+_-48); x=_sign?-x:x;}while(0)
#define sc(x) do{while((x=getchar())<33);}while(0)
char _; bool _sign;
// clang-format on

const int MM = 2e5 + 3;
int n, r, x, y, ans = INT_MAX, dis[MM], par[MM];
vector<int> adj[MM];
deque<int> q;
int main() {
    su(n);
    su(r);
    for (int i = 0, a, b; i < n - 1; i++) {
        su(a);
        su(b);
        adj[a].push_back(b);
        adj[b].push_back(a);
    }
    memset(dis, -1, sizeof(dis));
    for (int i = 0, a; i < r; i++) {
        su(a);
        dis[a] = 0;
        q.push_back(a);
    }
    while (!q.empty()) {
        int cur = q.front();
        q.pop_front();
        for (int nxt : adj[cur]) {
            if (dis[nxt] == -1) {
                dis[nxt] = dis[cur] + 1;
                q.push_back(nxt);
            }
        }
    }
    su(x);
    su(y);
    q.push_back(x);
    par[x] = -1;
    while (!q.empty()) {
        int cur = q.front();
        q.pop_front();
        for (int nxt : adj[cur]) {
            if (par[nxt] == 0) {
                par[nxt] = cur;
                q.push_back(nxt);
            }
        }
    }
    for (; y != -1; y = par[y]) {
        ans = min(ans, dis[y]);
    }
    printf("%d\n", ans);

    return 0;
}
